File: gameOfLife/Grid.py
from Cell import Cell

class Grid(object) :

    #construct a grid of cells
    def __init__(self, numRows, numColumns) :
        self.numRows = numRows
        self.numColumns = numColumns
        self.itsCurrentState = [ [ Cell() for x in range(numRows)] for y in range(numColumns)]
        self.itsNextState =    [ [ Cell() for x in range(numRows)] for y in range(numColumns)]

    def cellIsAlive(self, x, y) :
        return (self.itsCurrentState[x][y]).isAlive()

    def update(self) :
        # loop over rows
        for row in range(self.numRows) :
            for column in range(self.numColumns) :
                isAliveNextRound = self.aliveNextRound(row,column)
                self.itsNextState[row][column].setAlive(isAliveNextRound)
        self.swapStates();


    def aliveNextRound(self, row, column) :
        aliveNextRound = False
        currentAliveState = (self.itsCurrentState[row][column]).isAlive()
        liveNeighbors = self.getCountOfLiveNeighbors(row,column)
        if currentAliveState == True and liveNeighbors <2 :
            aliveNextRound = False
        elif currentAliveState == True and (liveNeighbors == 2 or liveNeighbors == 3) :
            aliveNextRound = True
        elif  currentAliveState == True and liveNeighbors > 3 :
            aliveNextRound = False
        elif currentAliveState == False and liveNeighbors == 3 :
            aliveNextRound = True
        return aliveNextRound

    def getCountOfLiveNeighbors(self, row, column) :
        numLiveNeighbors = 0
        up = row-1
        down = row+1
        right = column+1
        left = column-1
        if row == 0 :  # top edge case
            up = self.numRows-1
        elif row == self.numRows-1 : # bottom edge case
            down = 0

        if column == 0 :  # left edge case
            left = self.numColumns-1
        elif column == self.numColumns-1 : # right edge case
            right = 0

        neighborsToConsider = [ [up,left],   [up,column], [up,right],
                                [row,left],               [row,right],
                                [down,left],[down,column],[down,right] ]

        for neighbor in neighborsToConsider :
            if self.itsCurrentState[neighbor[0]][neighbor[1]].isAlive() :
                numLiveNeighbors = numLiveNeighbors + 1

        return numLiveNeighbors;

    # this swaps current and next references
    def swapStates(self) :
        temp = self.itsCurrentState
        self.itsCurrentState = self.itsNextState
        self.itsNextState = temp
    
    def __str__(self) :
        returnVal = "";
        for row in range(self.numRows) :
            for column in range(self.numColumns) :
                returnVal += str(self.itsCurrentState[row][column])
                returnVal += " "
            returnVal += "\n"
        return returnVal

    def gliderSetup(self) :
			self.itsCurrentState[5][5].setAlive(True)
			self.itsCurrentState[6][5].setAlive(True)
			self.itsCurrentState[7][5].setAlive(True)
			self.itsCurrentState[7][4].setAlive(True)
			self.itsCurrentState[6][3].setAlive(True)

File: GameOfLifeJava/Grid.java
import java.util.Arrays;


public class Grid {
	private int numOfRows = 0;
	private int numOfColumns = 0;
	public 	Cell[][] currentState ;
	private Cell[][] nextState ;
	
	
	public Grid(int numOfRows, int numOfColumns){
		this.numOfRows = numOfRows;
		this.numOfColumns = numOfColumns;
		this.currentState = new Cell[numOfRows][numOfColumns];
		this.nextState = new Cell[numOfRows][numOfColumns];
		for (int i = 0; i < numOfRows; i++){
			for (int j = 0; j < numOfColumns; j++){
				currentState[i][j] = new Cell();
				nextState[i][j] = new Cell();
			}
			
		}
	}
	
	public void GliderSetup(){
		this.currentState[5][5].setAlive(true);
		this.currentState[6][5].setAlive(true);
		this.currentState[7][5].setAlive(true);
		this.currentState[7][4].setAlive(true);
        this.currentState[6][3].setAlive(true);
		
	}
	
	public int getNeighbourCount(int row, int column){
		int numberOfNeighbours = 0;
			int up = row - 1 ;
			int down = row + 1;
			int left = column - 1;
			int right = column + 1;
			if (row == 0){
				up = this.numOfRows-1;
			}
			if (row == (this.numOfRows-1)){
				down = 0;
			}
			if (column == (this.numOfColumns-1)){
				right = 0;
			}
			if (column == 0){
				left = this.numOfColumns-1;
			}
			
			int[][] neighbours = {{up, left}, {up, column}, {up, right}, {row, left}, {row, right}, {down, left}, {down, column}, {down, right}};
			
			for (int[] j : neighbours){
				if (this.currentState[j[0]][j[1]].isAlive()){
					numberOfNeighbours++;
				}
			}
			
		return numberOfNeighbours;
		
	}

	public boolean isAliveNextRound(int row, int column){
		boolean aliveNextRound = false;
			boolean current = currentState[row][column].isAlive();
			int liveNeighbours = getNeighbourCount(row, column);
			if (current == true && liveNeighbours < 2){
				aliveNextRound = false;
			}else if (current == true && (liveNeighbours == 2 || liveNeighbours == 3)){
				aliveNextRound = true;
			}else if (current == false && liveNeighbours == 3){
				aliveNextRound = true;
			}else {
				aliveNextRound = false;
			}
			
			return aliveNextRound;
	}

	public void update(){
		for (int i = 0; i < this.numOfRows; i ++){
			for (int j = 0; j < numOfColumns; j++){
				boolean alivenext=  isAliveNextRound(i, j);
				this.nextState[i][j].setAlive(alivenext);
			}
		}
		
		swapStates();
	
	}

	private void swapStates(){
		Cell[][] temp = this.currentState;
		this.currentState = this.nextState;
		this.nextState = temp;
		
	}


	

}
